File: tests/leetcode/x/ruby/0317.rb
def shortest_distance(grid)
  rows = grid.length
  cols = grid[0].length
  dist = Array.new(rows) { Array.new(cols, 0) }
  reach = Array.new(rows) { Array.new(cols, 0) }
  buildings = 0

  rows.times do |sr|
    cols.times do |sc|
      next unless grid[sr][sc] == 1

      buildings += 1
      seen = Array.new(rows) { Array.new(cols, false) }
      q = [[sr, sc, 0]]
      seen[sr][sc] = true
      head = 0
      while head < q.length
        r, c, d = q[head]
        head += 1
        [[1, 0], [-1, 0], [0, 1], [0, -1]].each do |dr, dc|
          nr = r + dr
          nc = c + dc
          next unless nr.between?(0, rows - 1) && nc.between?(0, cols - 1)
          next if seen[nr][nc]

          seen[nr][nc] = true
          if grid[nr][nc] == 0
            dist[nr][nc] += d + 1
            reach[nr][nc] += 1
            q << [nr, nc, d + 1]
          end
        end
      end
    end
  end

  ans = nil
  rows.times do |r|
    cols.times do |c|
      next unless grid[r][c] == 0 && reach[r][c] == buildings

      ans = ans.nil? ? dist[r][c] : [ans, dist[r][c]].min
    end
  end
  ans || -1
end

data = STDIN.read.split.map(&:to_i)
exit if data.empty?

pos = 0
t = data[pos]
pos += 1
blocks = []
t.times do
  rows = data[pos]
  cols = data[pos + 1]
  pos += 2
  grid = Array.new(rows) do
    row = data[pos, cols]
    pos += cols
    row
  end
  blocks << shortest_distance(grid).to_s
end

print blocks.join("\n\n")

File: tests/leetcode/x/cpp/0317.cpp
#include <iostream>
#include <queue>
#include <vector>

using namespace std;

int shortestDistance(const vector<vector<int>>& grid) {
    int rows = grid.size(), cols = grid[0].size();
    vector<vector<int>> dist(rows, vector<int>(cols, 0));
    vector<vector<int>> reach(rows, vector<int>(cols, 0));
    int buildings = 0;
    for (int sr = 0; sr < rows; ++sr) {
        for (int sc = 0; sc < cols; ++sc) {
            if (grid[sr][sc] != 1) continue;
            buildings++;
            vector<vector<bool>> seen(rows, vector<bool>(cols, false));
            queue<tuple<int, int, int>> q;
            q.push({sr, sc, 0});
            seen[sr][sc] = true;
            while (!q.empty()) {
                auto [r, c, d] = q.front();
                q.pop();
                const int dr[4] = {1, -1, 0, 0};
                const int dc[4] = {0, 0, 1, -1};
                for (int k = 0; k < 4; ++k) {
                    int nr = r + dr[k], nc = c + dc[k];
                    if (nr >= 0 && nr < rows && nc >= 0 && nc < cols && !seen[nr][nc]) {
                        seen[nr][nc] = true;
                        if (grid[nr][nc] == 0) {
                            dist[nr][nc] += d + 1;
                            reach[nr][nc] += 1;
                            q.push({nr, nc, d + 1});
                        }
                    }
                }
            }
        }
    }
    int ans = -1;
    for (int r = 0; r < rows; ++r) {
        for (int c = 0; c < cols; ++c) {
            if (grid[r][c] == 0 && reach[r][c] == buildings) {
                if (ans == -1 || dist[r][c] < ans) ans = dist[r][c];
            }
        }
    }
    return ans;
}

int main() {
    int t;
    if (!(cin >> t)) return 0;
    for (int tc = 0; tc < t; ++tc) {
        int r, c;
        cin >> r >> c;
        vector<vector<int>> grid(r, vector<int>(c));
        for (int i = 0; i < r; ++i)
            for (int j = 0; j < c; ++j)
                cin >> grid[i][j];
        if (tc) cout << "\n\n";
        cout << shortestDistance(grid);
    }
    return 0;
}
